### PR TITLE
Usage telemetry infrastructure

### PR DESCRIPTION
**Pull request by richardm-stripe**
Contains the parts of https://github.com/stripe/stripe-go/pull/1779 that can be included in the master branch.

### CODE REVIEW

File: stripe_test.go [hunk 1]
```diff
@@ -560,25 +546,27 @@ func TestDo_TelemetryEnabled(t *testing.T) {
...
 
 		assert.NoError(t, err)
 		assert.Equal(t, message, response.Message)
 	}
 
 	// We should have seen exactly two requests.
 	assert.Equal(t, 2, requestNum)
+	// The telemetry in the second request should contain the
+	// expected usage
+	assert.Equal(t, telemetry.LastRequestMetrics.Usage, []string{"llama", "bufo"})
```
Comment (pakrym-stripe) [inline stripe_test.go:569]: Why not validate next to the RequestDurationMS validation?

**Review by pakrym-stripe — APPROVED** [review]

File: stripe_test.go (hunk 1 above)
Comment (richardm-stripe) [reply stripe_test.go]: Fixed!

File: stripe_test.go [hunk 2]
```diff
@@ -496,24 +487,20 @@ func TestDo_TelemetryDisabled(t *testing.T) {
...
-	}
-
 	type requestTelemetry struct {
 		LastRequestMetrics requestMetrics `json:"last_request_metrics"`
 	}
 
 	message := "Hello, client."
 	requestNum := 0
 
+	var telemetry requestTelemetry
```
Comment (pakrym-stripe) [inline stripe_test.go:503]: nit: does this variable have to stay top level? or can it move back into the server callback?

Comment (richardm-stripe) [reply stripe_test.go:503]: Will do!

**Pull request merged**